File: 3k-1s/architecture/laba4/benchit/kernel/memory/bandwidth/F77/OpenMP/0/double_stream/work.f
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
C BenchIT - Performance Measurement for Scientific Applications
C Contact: developer@benchit.org
C
C $Id: work.f 1 2009-09-11 12:26:19Z william $
C $URL: svn+ssh://william@rupert.zih.tu-dresden.de/svn-base/benchit-root/BenchITv6/kernel/memory/bandwidth/F77/OpenMP/0/double_stream/work.f $
C For license details see COPYING in the package base directory
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
C Kernel: measure Bandwidth inspired by STREAM benchmark
C         (FORTRAN OMP-version)
C
C         according to the rules, reffer this Benchmark as:
C         "BenchIT kernel based on a variant of the STREAM benchmark code"
C         when publishing results
C
C         This file contains the work, that is done: copy,scale,add
C         and triad
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC

	SUBROUTINE copy( c ,a , n)
	INTEGER*4 n,i
	double precision a(n)
	double precision c(n)
!$omp parallel do
	do 100 i=1,n
	   c(i)=a(i)
 100	CONTINUE
	RETURN
	END
	
	SUBROUTINE scale( b, c, scalar, n)
	INTEGER*4 n,i
	double precision b(n)
	double precision c(n)
	double precision scalar
!$omp parallel do
	do 100 i=1,n
	   b(i)=scalar*c(i)
 100	CONTINUE
	RETURN
	END
	
	SUBROUTINE add( c, a, b, n)
	INTEGER*4 n,i
	double precision a(n)
	double precision b(n)
	double precision c(n)
!$omp parallel do
	do 100 i=1,n
	   c(i)=a(i)+b(i)
 100	CONTINUE
	RETURN
	END
	
	SUBROUTINE triad( a, b, c, scalar, n)
	INTEGER*4 n,i
	double precision a(n)
	double precision b(n)
	double precision c(n)
	double precision scalar
!$omp parallel do
	do 100 i=1,n
	   a(i)=b(i)+scalar*c(i)
 100	CONTINUE
	RETURN
	END

File: 3k-1s/architecture/laba4/benchit/kernel/memory/bandwidth/C/OpenMP/0/double_stream/work.c
/********************************************************************
 * BenchIT - Performance Measurement for Scientific Applications
 * Contact: developer@benchit.org
 *
 * $Id: work.c 1 2009-09-11 12:26:19Z william $
 * $URL: svn+ssh://william@rupert.zih.tu-dresden.de/svn-base/benchit-root/BenchITv6/kernel/memory/bandwidth/C/OpenMP/0/double_stream/work.c $
 * For license details see COPYING in the package base directory
 *******************************************************************/
/* Kernel: measure Bandwidth inspired by STREAM benchmark (C OMP-version)
 *
 *         according to the rules, reffer this Benchmark as:
 *         "BenchIT kernel based on a variant of the STREAM benchmark code"
 *         when publishing results.
 *
 *         This file contains the work, that is done: copy,scale,add
 *         and triad
 *******************************************************************/

#include "work.h" 
 
void copy_(double *a, double *b, int size)
{
	register int i=0;
	#pragma omp parallel for
	for (i=0;i<size;i++)
	{
		a[i]=b[i];
	}
}
void scale_(double *a, double *b, double scalar, int size)
{
	register int i=0;
	#pragma omp parallel for
	for (i=0;i<size;i++)
	{
		a[i]=scalar*b[i];
	}
}
void add_(double *a, double *b, double *c, int size)
{
	register int i=0;
	#pragma omp parallel for
	for (i=0;i<size;i++)
	{
		a[i]=b[i]+c[i];
	}
}
void triad_(double *a, double *b, double *c, double scalar, int size)
{
	register int i=0;
	#pragma omp parallel for
	for (i=0;i<size;i++)
	{
		a[i]=b[i]+scalar*c[i];
	}
}
 

